packages/@aws-cdk/cdk/test/core/test.token-json.ts

```diff
@@ -0,0 +1,175 @@
+import { Test } from 'nodeunit';
+import { CloudFormationIntrinsicToken, FnConcat, resolve, Token, TokenJSON } from '../../lib';
+import { evaluateCFN } from './evaluate-cfn';
+
+export = {
+    'plain JSON.stringify() on a Token fails'(test: Test) {
+        // GIVEN
+        const token = new Token(() => 'value');
+
+        // WHEN
+        test.throws(() => {
+            JSON.stringify({ token });
+        });
+
+        test.done();
+    },
+
+    'string tokens can be JSONified and JSONification can be reversed'(test: Test) {
+        for (const token of tokensThatResolveTo('woof woof')) {
+            // GIVEN
+            const fido = { name: 'Fido', speaks: token };
+
+            // WHEN
+            const resolved = resolve(TokenJSON.stringify(fido));
+
+            // THEN
+            test.deepEqual(evaluateCFN(resolved), '{"name":"Fido","speaks":"woof woof"}');
+        }
+
+        test.done();
+    },
+
+    'string tokens can be embedded while being JSONified'(test: Test) {
+        for (const token of tokensThatResolveTo('woof woof')) {
+            // GIVEN
+            const fido = { name: 'Fido', speaks: `deep ${token}` };
+
+            // WHEN
+            const resolved = resolve(TokenJSON.stringify(fido));
+
+            // THEN
+            test.deepEqual(evaluateCFN(resolved), '{"name":"Fido","speaks":"deep woof woof"}');
+        }
+
+        test.done();
+    },
+
+    'integer Tokens behave correctly in stringification and JSONification'(test: Test) {
+        // GIVEN
+        const num = new Token(() => 1);
+        const embedded = `the number is ${num}`;
+
+        // WHEN
+        test.equal(evaluateCFN(resolve(embedded)), "the number is 1");
+        test.equal(evaluateCFN(resolve(TokenJSON.stringify({ embedded }))), "{\"embedded\":\"the number is 1\"}");
+        test.equal(evaluateCFN(resolve(TokenJSON.stringify({ num }))), "{\"num\":1}");
+
+        test.done();
+    },
+
+    'tokens in strings survive additional TokenJSON.stringification()'(test: Test) {
+        // GIVEN
+        for (const token of tokensThatResolveTo('pong!')) {
+            // WHEN
+            const stringified = TokenJSON.stringify(`ping? ${token}`);
+
+            // THEN
+            test.equal(evaluateCFN(resolve(stringified)), '"ping? pong!"');
+        }
+
+        test.done();
+    },
+
+    'intrinsic Tokens embed correctly in JSONification'(test: Test) {
+        // GIVEN
+        const bucketName = new CloudFormationIntrinsicToken({ Ref: 'MyBucket' });
+
+        // WHEN
+        const resolved = resolve(TokenJSON.stringify({ theBucket: bucketName }));
+
+        // THEN
+        const context = {MyBucket: 'TheName'};
+        test.equal(evaluateCFN(resolved, context), '{"theBucket":"TheName"}');
```

Comment:
The reason why we have this: I actually don't care about the CFN specifics. Use either `{Fn::Join}` or `{Fn::Sub}` or anything else--as long as it evaluates to whatever I care about. If we decide to change the mechanism that is used, I expect these tests to still work, because the feature still works.

If this is an issue of trust in the specific implementation of `evaluateCFN`, that is remedied with either tests on the implementation of it (but seems like overkill to me), or deploying a couple of stacks that use the current Token stringification implementation and seeing that they work (and thereby increasing the trust that `evaluateCFN` does what CloudFormation does), which is done in the integ test of CloudWatch dashboards.

Reply:
Case in point: I just refactor the code to combine string literals while evaluating, so that if there were 2 fragments it was guaranteed that one of them was an intrinsic, and the only test that broke was one that did NOT use `evaluateCFN`. The test that broke was the one that looked at the structure of the returned data, not the thing it was representing (which is what is interesting about it).

What I'm concerned with here is testing at the right level of concern: I have a feeling that many of our tests that are testing "output is exactly this" are testing too much, not testing the actual behavior that they're interested in. This means they are less readable and prone to breakage. That, in turn, leads to more time to make changes (because you need to fix all tests that now complain) and also change fatigue, where after a while I will either (a) stop making changes or (b) ignore the cause of the failing test, assume it's because of something trivial and just copy/paste in a new example. Both would be bad.

Reply:
:-)

I understand what you are talking about. It is a pragmatic choice we have been making throughout the CDK (and jsii to be honest). I think it's a cost/benefit question. It's much much easier to write those types of heavy-handed "diff" tests then granular, pure, meticulous testing you are talking about. Of course those would be more precise, but since we are still at a stage where things are moving quite a lot, the efficiency of using "diff tests" is proving itself. It might be annoying to need to update the test expectations, but it's much simpler to do that then to start changing dozens of potential assertions.

It's also about tools. I think our assertion library started providing some tools such as `expect(to(haveResource(` that make it easier to author more precise tests. But in a sense, we might be missing things because it's not testing for === but rather for >=. 

At any rate, as much as this is unpure, I am quite okay with this trade off so far. It's proven itself quite good in protecting us from regression, which is, at the end, the sole purpose of those tests.

Reply:
I know and I agree.

But the `evaluateCFN` function is both a really rather low investment AND allows us to express tests at the right level of abstraction ("if the result of this were to be evaluated by CloudFormation, would it give us back the value we intended?").

Reply:
I subscribe to the "one concern per test" philosophy, so that a test has one reason to break (because that concern is no longer satisfied).

If you really really care, I will add a "String concatenation required by Token stringification is implemented using `{Fn::Join}`" test, so that we can guarantee that `{Fn::Join}` is used. Then, if we ever change that to `{Fn::Sub}`, that new test will rightly break because it was asserting that we were using `{Fn::Join}` and that's no longer true.

But I'm not going to rewrite the other tests to make the same assertions, because they're testing different concerns (embedding and escaping and such).

Reply:
Sure